### PR TITLE
Fix Drop Function validation to validate sub-columns of objects

### PR DESCRIPTION
**Pull request by jeeminso**
## Summary of the changes / Why this improves CrateDB
Fixes https://github.com/crate/crate/issues/13638

## Checklist

 - [x] Added an entry in `CHANGES.txt` for user facing changes
 - [x] Updated documentation & `sql_features` table for user facing changes
 - [x] Touched code is covered by tests
 - [x] [CLA](https://crate.io/community/contribute/cla/) is signed
 - [x] This does not contain breaking changes, or if it does:
    - It is released within a major release
    - It is recorded in ``CHANGES.txt``
    - It was marked as deprecated in an earlier release if possible
    - You've thought about the consequences and other components are adapted
      (E.g. AdminUI)

### CODE REVIEW

File: server/src/test/java/io/crate/expression/udf/UserDefinedFunctionServiceTest.java [hunk 1]
```diff
@@ -158,4 +159,22 @@ public void test_validate_partitioned_table_while_dropping_udf() throws Exceptio
...
+            .addPartitionedTable("create table doc.p1 (o object as (id int), gen as foo(o['id'])) partitioned by (o['id'])")
+            .build();
+
+        assertThatThrownBy(() -> executor.udfService().validateFunctionIsNotInUseByGeneratedColumn(
+                               Schemas.DOC_SCHEMA_NAME,
+                               "foo",
+                               metadataWithoutFunction,
+                               clusterService.state()))
+            .hasMessageContaining("Cannot drop function 'foo', it is still in use by 'doc.p1.gen AS doc.foo(o['id'])'")
+            .isInstanceOf(IllegalArgumentException.class);
```
Comment (matriv) [inline server/src/test/java/io/crate/expression/udf/UserDefinedFunctionServiceTest.java:178]:
```suggestion
            .isExactlyInstanceOf(IllegalArgumentException.class);
```
and could you please move it one line above before asserting the message (just for consistency)

File: server/src/main/java/io/crate/expression/udf/UserDefinedFunctionService.java [hunk 2]
```diff
@@ -301,7 +301,11 @@ void validateFunctionIsNotInUseByGeneratedColumn(String schema,
 
         for (var indexParts : indices) {
             var tableInfo = docTableFactory.create(indexParts.toRelationName(), currentState);
-            TableReferenceResolver tableReferenceResolver = new TableReferenceResolver(tableInfo.columns(), tableInfo.ident());
+            var functionParameters = tableInfo.generatedColumns().stream()
```
Comment (matriv) [inline server/src/main/java/io/crate/expression/udf/UserDefinedFunctionService.java:304]: Not a big fan of the streams usage, I'd personally convert to a for loop, but also not strong opinion.

**Review by matriv — APPROVED** [review]
LGTM. Left a couple of comments, thank you!

**Comment by jeeminso** [comment]
Thanks for reviewing @matriv ! Do you think we need changes doc entry for this? This is only exception type and message changes.

**Comment by matriv** [comment]
> Thanks for reviewing @matriv ! Do you think we need changes doc entry for this? This is only exception type and message changes.

Why not? It's still a fix, since the original message was misleading.

File: server/src/test/java/io/crate/expression/udf/UserDefinedFunctionServiceTest.java (hunk 1 above)
Comment (jeeminso) [reply server/src/test/java/io/crate/expression/udf/UserDefinedFunctionServiceTest.java]: Done, Thank you

File: server/src/main/java/io/crate/expression/udf/UserDefinedFunctionService.java (hunk 2 above)
Comment (jeeminso) [reply server/src/main/java/io/crate/expression/udf/UserDefinedFunctionService.java]: Also done, Thanks!

**Pull request merged**